File: app/src/main/java/fragment/ListaSeguindoFragment.java
package fragment;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.crashlytics.android.Crashlytics;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import adapter.ProximosAdapter;
import adapter.SeguindoRecycleAdapter;
import br.com.icaro.filme.R;
import domain.UserEp;
import domain.UserSeasons;
import domain.UserTvshow;
import utils.Constantes;

/**
 * Created by icaro on 25/11/16.
 */
public class ListaSeguindoFragment extends Fragment {

    private final String TAG = ListaSeguindoFragment.class.getName();
    private List<UserTvshow> userTvshows;
    private int tipo;
    private RecyclerView recyclerViewMissing;
    private RecyclerView recyclerViewSeguindo;
    private ValueEventListener eventListener;
    private DatabaseReference seguindoDataBase;

    public static Fragment newInstance(int tipo, List<UserTvshow> userTvshows) {
        ListaSeguindoFragment fragment = new ListaSeguindoFragment();
        Bundle bundle = new Bundle();
        bundle.putSerializable(Constantes.INSTANCE.getSEGUINDO(), (Serializable) userTvshows);
        bundle.putInt(Constantes.INSTANCE.getABA(), tipo);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            tipo = getArguments().getInt(Constantes.INSTANCE.getABA());
            userTvshows = (List<UserTvshow>) getArguments().getSerializable(Constantes.INSTANCE.getSEGUINDO());
        }

        FirebaseAuth mAuth = FirebaseAuth.getInstance();
        FirebaseDatabase database = FirebaseDatabase.getInstance();
        seguindoDataBase = database.getReference("users").child(mAuth.getCurrentUser()
                .getUid()).child("seguindo");
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        switch (tipo) {

            case 0: {
                return getViewMissing(inflater, container);

            }
            case 1: {
                return getViewSeguindo(inflater, container);
            }
        }
        return null;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        eventListener = new ValueEventListener() {
            @Override
            public void onDataChange(DataSnapshot dataSnapshot) {
                userTvshows = new ArrayList<>();
                if (dataSnapshot.exists()) {

                    for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                        try {
                            UserTvshow userTvshow = snapshot.getValue(UserTvshow.class);
                            userTvshows.add(userTvshow);
                        } catch (Exception e) {
                            Crashlytics.logException(e);
                            if (snapshot.hasChild("nome") && getActivity() != null) {
                                String nome = snapshot.child("nome").getValue(String.class);
                                Toast.makeText(getActivity(), getResources().getString(R.string.ops_seguir_novamente) + " - " + nome, Toast.LENGTH_LONG).show();
                            } else {
                                if (getActivity() != null) {
                                    Toast.makeText(getActivity(), getResources().getString(R.string.ops_seguir_novamente), Toast.LENGTH_LONG).show();
                                }
                            }
                        }
                    }
                    if (getView() != null) {
                        recyclerViewMissing = (RecyclerView) getView().getRootView().findViewById(R.id.temporadas_recycle);
                        recyclerViewSeguindo = (RecyclerView) getView().getRootView().findViewById(R.id.seguindo_recycle);
                        recyclerViewMissing.setAdapter(new ProximosAdapter(getActivity(), setSeriesMissing(userTvshows)));
                        recyclerViewSeguindo.setAdapter(new SeguindoRecycleAdapter(getActivity(), userTvshows));
                    }
                }
            }

            @Override
            public void onCancelled(DatabaseError databaseError) {

            }
        };
        seguindoDataBase.addValueEventListener(eventListener);

    }

    private List<UserTvshow> setSeriesMissing(List<UserTvshow> userTvshows) {
        List<UserTvshow> temp = new ArrayList<>();

        for (UserTvshow userTvshow : userTvshows) {
            boolean season = true;
            for (UserSeasons seasons : userTvshow.getSeasons()) {
                if (seasons.getSeasonNumber() != 0 && seasons.getUserEps() != null && season)
                    for (UserEp userEp : seasons.getUserEps()) {
                        if (!userEp.isAssistido()) {
                            temp.add(userTvshow);
                            season = false;
                            break;
                        }
                    }
            }
        }// gambiara. Arrumar!

        return temp;
    }

    private View getViewMissing(LayoutInflater inflater, ViewGroup container) {
        View view = inflater.inflate(R.layout.temporadas, container, false); // Criar novo layout
        view.findViewById(R.id.progressBarTemporadas).setVisibility(View.GONE);
        recyclerViewMissing = (RecyclerView) view.findViewById(R.id.temporadas_recycle);
        recyclerViewMissing.setHasFixedSize(true);
        recyclerViewMissing.setItemAnimator(new DefaultItemAnimator());
        recyclerViewMissing.setLayoutManager(new LinearLayoutManager(getContext()));
        List<UserTvshow> missing = setSeriesMissing(userTvshows);
        if (missing.size() > 0) {
            recyclerViewMissing.setAdapter(new ProximosAdapter(getActivity(), missing));
        } else {
            view.findViewById(R.id.text_search_empty).setVisibility(View.VISIBLE);
            ((TextView) view.findViewById(R.id.text_search_empty)).setText(R.string.empty);
        }
        return view;
    }

    private View getViewSeguindo(LayoutInflater inflater, ViewGroup container) {
        View view = inflater.inflate(R.layout.seguindo, container, false); // Criar novo layout
        view.findViewById(R.id.progressBarTemporadas).setVisibility(View.GONE);
        recyclerViewSeguindo = (RecyclerView) view.findViewById(R.id.seguindo_recycle);
        recyclerViewSeguindo.setHasFixedSize(true);
        recyclerViewSeguindo.setItemAnimator(new DefaultItemAnimator());
        recyclerViewSeguindo.setLayoutManager(new GridLayoutManager(getContext(), 4));
        if (userTvshows.size() > 0) {
            recyclerViewSeguindo.setAdapter(new SeguindoRecycleAdapter(getActivity(), userTvshows));
        } else {
            view.findViewById(R.id.text_search_empty).setVisibility(View.VISIBLE);
            ((TextView) view.findViewById(R.id.text_search_empty)).setText(R.string.empty);
        }
        return view;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (eventListener != null) {
            seguindoDataBase.removeEventListener(eventListener);
        }
    }
}

File: app/src/main/java/fragment/ListaSeguindoFragment.kt
package fragment

import adapter.ProximosAdapter
import adapter.SeguindoRecycleAdapter
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import br.com.icaro.filme.R
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.FirebaseDatabase
import domain.Api
import domain.UserTvshow
import domain.tvshow.Tvshow
import kotlinx.coroutines.experimental.*
import utils.Constantes
import utils.UtilsApp
import java.io.Serializable
import java.util.*

/**
 * Created by icaro on 25/11/16.
 */
class ListaSeguindoFragment : Fragment() {

    private var userTvshows: MutableList<UserTvshow>? = null
    private var tipo: Int = 0
    private var recyclerViewMissing: RecyclerView? = null
    private var recyclerViewSeguindo: RecyclerView? = null
    private var rotina: Job? = null
    private var adapterProximo: ProximosAdapter? = null
    private var adapterSeguindo: SeguindoRecycleAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            tipo = arguments!!.getInt(Constantes.ABA)
            userTvshows = arguments!!.getSerializable(Constantes.SEGUINDO) as MutableList<UserTvshow>
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        when (tipo) {

            0 -> {
                return getViewMissing(inflater, container)

            }
            1 -> {
                return getViewSeguindo(inflater, container)
            }
        }
        return null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d("Icaro", "onViewCreated")
        when (tipo) {
            0 -> {
                //verificarProximoEp()
            }

            1 -> {
                verificarSerieCoroutine()
            }
        }
    }

    fun verificarSerieCoroutine() {
        userTvshows?.forEachIndexed { indexFire, tvFire ->
            try {
                rotina = GlobalScope.launch(Dispatchers.Main) {
                    val serie = async(Dispatchers.IO) { Api(context = context!!).getTvShowLiteC(tvFire.id) }.await()
                    if (serie.id == null) {
                        adapterSeguindo?.add(tvFire)
                        return@launch
                    }
                    if (serie.numberOfEpisodes != tvFire.numberOfEpisodes) {
                        tvFire.desatualizada = true
                        adapterSeguindo?.add(tvFire)

                    } else {
                        adapterSeguindo?.add(tvFire)
                    }
                }
            } catch (ex: Exception) {
                ex.message
            }
        }
    }

    private fun getViewMissing(inflater: LayoutInflater, container: ViewGroup?): View {
        Log.d("Icaro", "getViewMissing")
        val view = inflater.inflate(R.layout.temporadas, container, false) // Criar novo layout
        view.findViewById<View>(R.id.progressBarTemporadas).visibility = View.GONE
        adapterProximo = ProximosAdapter(this.activity, mutableListOf<UserTvshow>())
        recyclerViewMissing = view.findViewById<View>(R.id.temporadas_recycle) as RecyclerView
        recyclerViewMissing!!.setHasFixedSize(true)
        recyclerViewMissing!!.itemAnimator = DefaultItemAnimator()
        recyclerViewMissing!!.layoutManager = LinearLayoutManager(context)
        recyclerViewMissing!!.adapter = adapterProximo
        return view
    }

    private fun getViewSeguindo(inflater: LayoutInflater, container: ViewGroup?): View {
        val view = inflater.inflate(R.layout.seguindo, container, false) // Criar novo layout
        view.findViewById<View>(R.id.progressBarTemporadas).visibility = View.GONE
        recyclerViewSeguindo = view.findViewById<View>(R.id.seguindo_recycle) as RecyclerView
        adapterSeguindo = SeguindoRecycleAdapter(activity, mutableListOf<UserTvshow>())
        recyclerViewSeguindo!!.setHasFixedSize(true)
        recyclerViewSeguindo!!.itemAnimator = DefaultItemAnimator()
        recyclerViewSeguindo!!.layoutManager = GridLayoutManager(context, 4)
        recyclerViewSeguindo!!.adapter = adapterSeguindo
        return view
    }

    override fun onDestroy() {
        super.onDestroy()
        if (rotina != null) {
            rotina!!.cancel()
        }
    }

    companion object {

        fun newInstance(tipo: Int, userTvshows: List<UserTvshow>): Fragment {
            val fragment = ListaSeguindoFragment()
            val bundle = Bundle()
            bundle.putSerializable(Constantes.SEGUINDO, userTvshows as Serializable)
            bundle.putInt(Constantes.ABA, tipo)
            fragment.arguments = bundle

            return fragment
        }
    }
}
